sdks/java/io/thrift/src/main/java/org/apache/beam/sdk/io/thrift/ThriftSchema.java

```diff
@@ -243,37 +253,65 @@ private FieldValueTypeInformation fieldValueTypeInfo(Class<?> type, String field
   @Override
   public @NonNull SchemaUserTypeCreator schemaTypeCreator(
       @NonNull Class<?> targetClass, @NonNull Schema schema) {
-    return params -> restoreThriftObject(targetClass, params);
+    return thriftMapper(targetClass, schema);
+  }
+
+  private <FieldT extends Enum<FieldT> & TFieldIdEnum, T extends TBase<T, FieldT>>
+      SchemaUserTypeCreator thriftMapper(Class<?> targetClass, Schema schema) {
+    final Map<FieldT, FieldMetaData> fieldDescriptors = schemaFieldDescriptors(targetClass, schema);
+    return params -> restoreThriftObject(targetClass, fieldDescriptors, params);
   }
 
+  @SuppressWarnings("nullness")
   private <FieldT extends Enum<FieldT> & TFieldIdEnum, T extends TBase<T, FieldT>>
-      T restoreThriftObject(Class<?> targetClass, Object[] params) {
+      Map<FieldT, FieldMetaData> schemaFieldDescriptors(Class<?> targetClass, Schema schema) {
```

Comment:
This allows using the order of fields in the incoming schema, instead of the original order in the thrift descriptor, so this will still work if fields are reordered by clients (the shuffled schema table test validates this).